### ISSUE TITLE
[Question] How to generate TOC data for custom React Component

### ISSUE THREAD

**Issue by max-pod** [hashicorp/next-mdx-remote#231]
Hi!

I'm trying to figure out how to generate TOC data to be filled in a custom react component.

I recognize that there are a couple of flexible TOC unified libraries, I think the most flexible is https://github.com/JS-DevTools/rehype-toc, which while it doesn't suit my use case perfectly, could probably be modified to work

Another solution I saw is to just run regexp on the content retrieved from matter() https://github.com/hashicorp/next-mdx-remote/issues/53#issuecomment-725906664

Something I would be really interested is in accessing the VFile data from https://github.com/stefanprobst/rehype-extract-toc, although I can't seem to figure out how to actually make it play with this library.

Any recommendations on how to proceed? I'm honestly a bit lost, although I really would like to use the rehype-extract-toc library.

While trying to understand the serialize function, I also ran into this:
```
function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }
var mdx__default = /*#__PURE__*/_interopDefaultLegacy(mdx);
```
For: https://github.com/hashicorp/next-mdx-remote/blob/acbcff81a445d20337996b490a15b7172d435c2f/src/serialize.ts#L77

What does _interopDefaultLegacy actually do? I get that it's typescript transforming ts into js, but I'm still confused.

TLDR: I think potential solutions for all into the categories of:
1. use a remark/rehype plugin to:
   a. explicitly write out a react component
   b. write out data a react component would use
2. use regexp

I would specifically favor 1b, although I'm hitting a pretty big roadblock. Would appreciate guidance.

Thank you for any help

**Comment by dekked** [comment]
Hey!

I wrote the following plugin that extracts headings with some rank and their slugs (`#id`). into a `headings` object you provide from the outside, you should be able to modify it to extract all.

```js
import has from "hast-util-has-property"
import headingRank from "hast-util-heading-rank"
import toString from "hast-util-to-string"
import visit from "unist-util-visit"

/**
 * Extracts headings from a unified tree.
 * 
 * To be used *AFTER* the `rehype-slug` plugin.
 *
 * Based on https://github.com/rehypejs/rehype-slug/blob/4.0.1/index.js
 */
export default function plugin({ rank = 2, headings }) {
  return function transformer(tree) {
    visit(tree, "element", function (node) {
      if (headingRank(node) === rank && has(node, "id")) {
        headings.push({
          title: toString(node),
          id: node.properties.id.toString(),
        })
      }
    })
  }
}
```

You would use it like:

```js
// ...
const headings = []

const mdxSource = await serialize(content, {
  scope: data,
  mdxOptions: {
    remarkPlugins: [], // whatever
    rehypePlugins: [
      // whatever
      [rehypeExtractHeadings, { rank: 2, headings }],
    ],
  },
})

// You got your headings available!
```

Note that the plugin would need to be updated for the newer ESM module dependencies. Here's how it should look like:
```js
import { hasProperty } from "hast-util-has-property"
import { headingRank } from "hast-util-heading-rank"
import { toString } from "hast-util-to-string"
import { visit } from "unist-util-visit"
import { Root } from "hast"

/**
* Extracts headings from a unified tree.
* 
* To be used *AFTER* the `rehype-slug` plugin.
*/
export default function rehypeExtractHeadings({
  rank = 2,
  headings,
}: ExtractHeadingsConfig) {
  return (tree: Root) => {
    visit(tree, "element", node => {
      if (
        headingRank(node) === rank &&
        node.properties &&
        hasProperty(node, "id")
      ) {
        headings.push({
          title: toString(node),
          id: node.properties.id.toString(),
        })
      }
    })
  }
}
```

Hope it's useful!

**Comment by marcelgruber** [comment]
Thank you @dekked.

What is `ExtractHeadingsConfig` in your example?

**Comment by marcelgruber** [comment]
@dekked I got your sample code working:

```javascript
import { hasProperty } from "hast-util-has-property"
import { headingRank } from "hast-util-heading-rank"
import toString from "hast-util-to-string"
import { visit } from "unist-util-visit"

const rehypeExtractHeadings = ({ rank = 2, headings }) => {
  return (tree) => {
    visit(tree, "element", function (node) {
      if (headingRank(node) === rank && hasProperty(node, "id")) {
        headings.push({
          title: toString(node),
          id: node.properties.id.toString()
        })
      }
    })
  }
}

export default rehypeExtractHeadings
```

**Comment by wangzishun** [comment]
you are my savior
